File: acorn-core/src/main/java/com/ardley/acorn/context/RequestContext.java
package com.ardley.acorn.context;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * A framework-agnostic representation of an incoming HTTP request's extractable context.
 *
 * <p>Provides access to path parameters, query parameters, and headers without coupling
 * extractors to any specific HTTP framework (JAX-RS, Spring MVC, Servlet API, etc.).
 *
 * <p>Framework integration modules are responsible for constructing this from their
 * native request types (e.g., {@code ContainerRequestContext} for JAX-RS,
 * {@code HttpServletRequest} for Spring).
 */
public interface RequestContext {

    /**
     * Returns the value of a named path parameter, or empty if not present.
     *
     * @param name the parameter name as declared in the route template (e.g., "id" from "/{id}")
     */
    Optional<String> pathParam(String name);

    /**
     * Returns the first value of a named query parameter, or empty if not present.
     *
     * @param name the query parameter name
     */
    Optional<String> queryParam(String name);

    /**
     * Returns all values of a named query parameter (for multi-valued params).
     *
     * @param name the query parameter name
     * @return list of values, empty list if not present
     */
    List<String> queryParams(String name);

    /**
     * Returns the first value of a named request header, or empty if not present.
     *
     * @param name the header name (case-insensitive)
     */
    Optional<String> header(String name);

    /**
     * Returns the full request path (e.g., "/users/u-123/documents").
     */
    String path();

    /**
     * Returns the HTTP method (GET, POST, PUT, DELETE, etc.).
     */
    String method();

    /**
     * Returns a named request property/attribute set by upstream middleware.
     * Used for passing data between filters.
     *
     * @param name the property name
     * @return the property value, or empty
     */
    Optional<Object> property(String name);
}

File: acorn-node/packages/express/src/request-context.ts
import type { Request } from 'express';
import type { RequestContext } from '@ardley/acorn-core';

export class ExpressRequestContext implements RequestContext {
  constructor(private readonly req: Request) {}

  pathParam(name: string): string | undefined {
    return this.req.params?.[name];
  }

  queryParam(name: string): string | undefined {
    const val = this.req.query?.[name];
    if (typeof val === 'string') return val;
    if (Array.isArray(val) && typeof val[0] === 'string') return val[0];
    return undefined;
  }

  queryParams(name: string): string[] {
    const val = this.req.query?.[name];
    if (typeof val === 'string') return [val];
    if (Array.isArray(val)) return val.filter((v): v is string => typeof v === 'string');
    return [];
  }

  header(name: string): string | undefined {
    const val = this.req.headers[name.toLowerCase()];
    if (typeof val === 'string') return val;
    if (Array.isArray(val)) return val[0];
    return undefined;
  }

  path(): string {
    return this.req.path;
  }

  method(): string {
    return this.req.method;
  }
}
